### PR TITLE
Implement the intermediate and hard level feature

### LINKED ISSUE #32: Implement level selection
Players should be able to select a level other than "Easy". Some ideas:

Harder:
- playing the tones at the same time
- limiting number of times they can replay the tones

Easier:
- give chances to use a "hint" function - e.g. to play the tones between the tones in a scale
- play tones only in an upward direction for easier levels (currently it plays sometimes in a downward direction)

### PR DESCRIPTION
**Pull request by eyeri91**
Closes #32 

# New features of `Intermediate` and `Hard` levels.

 ### 1. The number of hints available.

For the  `Intermediate` level, there will be 3 hints available and the `Hard` level won't get any hints.

Screen Shot 2022-10-19 at 4 51 56 PM

### 2. How to play 2 random notes 

**#playNotesDifferentlyDependingOnGameLeve(level)**

Those 2 random notes will play differently depending on the `#this.level`.

Screen Shot 2022-10-19 at 4 41 43 PM

If the current level is `Hard`, it will play 2 notes together to make it difficult to guess.

Screen Shot 2022-10-19 at 4 41 55 PM

# New event, `getHint` and `noHintAvailable`
 In the previous PR, `Implement Easy level feature`, I made a variable named `getHintButtomClickCounter` in `view.js`. Then I realized that limiting the number of hints has to do with the game logic. Therefore, I moved the idea to `game.js` and change its name to `numberOfHintsAvailable`. 

So when a user clicks the `getHintButton`, `view.js` will publish `this.#publishGetHintButtonClickedEvent`, which will update the `this.#numberOfHintsAvailable` in `game.js`  by calling `updateNumberOfHintsAvailable()`.

Screen Shot 2022-10-19 at 5 06 11 PM

Screen Shot 2022-10-19 at 4 47 56 PM

Screen Shot 2022-10-19 at 4 53 11 PM

⬆️  Once a user finishes up all hints he/she had, it will publish the `noHintAvailable` event,
which will disable the `getHintButton` ⬇️ 

Screen Shot 2022-10-19 at 4 53 35 PM

Screen Shot 2022-10-19 at 4 54 01 PM

Also, If the current level is `Hard`, it will publish the `noHintAvailable` event from the beginning of the game.

Screen Shot 2022-10-19 at 4 58 49 PM

-----------------------------------------------------------------------------------------------

I was not planning to implement both `Intermediate' and `Hard` levels at once but I was super excited to finally work on our project after back-to-back flights and just went on it. 😅 Because of too much excitement, I might have made lots of nonsense mistakes but let me know what you think of this approach!

Thank you ❤️

### CODE REVIEW

File: src/game.js
Comment (kuosandys) [inline src/game.js:21]:
```suggestion
```
Since we're not setting `this.#numberOfHintsAvailable` to a value, we can just declare it above - no need to do it here. Same with `this.#level`, we missed it before 😅

File: src/view.js
Comment (kuosandys) [inline src/view.js:23]: Nice improvement on making these private 👍

File: src/view.js
Comment (kuosandys) [inline src/view.js:72]:
```suggestion
    if (this.#appContainer.hasChildNodes()) {
      this.#appContainer.replaceChildren();
    }
```
There's no error here, but when `if` statements span multiple lines it's more readable to wrap the block of code to be executed in curly brackets `{}`

File: src/view.js
Comment (kuosandys) [inline src/view.js:213]:
```suggestion
  #playNotesForGameLevel(level) {
```
Small suggestion to make this a tiny bit more precise, but please feel free to ignore if you feel otherwise 😄

**Comment by kuosandys** [comment]
I really like the features you've implemented for the intermediate and hard levels! Identifying the interval with the two notes playing simultaneously will definitely take some serious skills 😄 

I've left some comments above, but overall the features work great. I see there are some merge conflicts to solve - I have a ll the confidence that you can figure it out, but let me know if I could help with anything there.

File: src/game.js
Comment (eyeri91) [reply src/game.js:21]: Aha! Nice detail!

File: src/view.js
Comment (eyeri91) [reply src/view.js:72]: I actually write down all the previous feedback and there was once a suggestion about removing curly brackets for one line arrow function. Ever since I was too obsessed with that and didn't consider the readability of code which is very important too. 😅 Thank you. ❤️

File: src/view.js
Comment (eyeri91) [reply src/view.js:213]: It clearly shows I had no good idea about naming this function haha. Thank you for your suggestion 😄

**Comment by eyeri91** [comment]
> I've left some comments above, but overall the features work great. I see there are some merge conflicts to solve - I have all the confidence that you can figure it out, but let me know if I could help with anything there.

I am working on merge conflicts at the moment and thought figured them out nicely. But apparently, it was not the case 😢 . I am getting this error below. Shall I git push my current changes in here or it would be a bad idea?

Screen Shot 2022-10-27 at 11 42 22 PM
Screen Shot 2022-10-27 at 11 43 07 PM

**Comment by eyeri91** [comment]
Sandy, I think I need your help in solving merging conflicts. One thing I would like to ask here is, as I was checking merge conflicts, it will play notes twice if we don't touch these parts of code. 


Here, it called `this.#playNotesForGameLevel(this.#level);` in `intermediate-level-feature` branch. If the level is `Easy` or `Intermediate` and I don't remove the lines below the `this.#playNotesForGameLevel(this.#level);` method, It will play notes twice. 
Screen Shot 2022-10-28 at 2 10 36 PM

Because `this.#playNotesForGameLevel(this.#level);` already has `this.#sampler.triggerAttackRelease....` lines. 

Screen Shot 2022-10-28 at 2 10 47 PM

In this case, is it okay if I trim the `this.#sampler.triggerAttackRelease.....` in the first pic? That part of code is written by you also, so I don't want to edit them without talking with you.

**Comment by eyeri91** [comment]
Another thing is running `git merge main` doesn't apply all new changes in the `main` branch to the `intermediate-level-feature` branch, So it is giving me some errors. In this case, should I leave them and sort them out when I git pull this PR and fix merge conflicts later or should I manually edit them in each file? It is hard to explain in typing 😅... I guess I am not helping you to understand my situation any better. 😢

File: src/view.js
Comment (kuosandys) [inline src/view.js:204]: Just tested it out after the changes were merged from main - I think we over-complicated this a bit previously, and this breaks the intermediate level.

Since we disable both buttons anyway when either are clicked, it's enough just to check whether the "Play tones" button is disabled.

Other than this I think we're good to merge!
```suggestion
            if (!playTonesButton.disabled) {
```

Comment (eyeri91) [reply src/view.js:204]: I read it through and you are right. I will update the file and gp.
By the way, I was just skimming through the `view.js` file, especially the CSS-related parts you wrote; your codes are so efficient and organized! I hope I could write like you soon ❤️

**Review by kuosandys — APPROVED** [review]
Awesome work, and really impressed with your determination to solve all the merge conflicts 😎 LGTM when you're ready!

**Pull request merged**
